Add initial docs (#21)
dapr-sandbox/components-go-sdk: daprdocs/content/en/go-sdk-docs/go-advanced/_index.md

---
type: docs
title: "Advanced uses of the Dapr pluggable components .Go SDK"
linkTitle: "Advanced"
weight: 2000

> description: How to use advanced techniques with with the Dapr pluggable components Go SDK

Author: Fixed.

is_preview: true
---

> While not typically needed by most, these guides show advanced ways to can configure your Go pluggable components.

Author: Fixed.

## Component lifetime

> Components are registered by passing a "factory method". This method will be called for each configured component of that type associated with that socket. The method returns the instance to associate with that component (whether shared or not). This allows multiple components of the same type to be configured with different sets of metadata, when component operations need to be isolated from one another, etc.

Reviewer: ```suggestion
Components are registered by passing a "factory method" that is called for each configured component type associated with that socket. The method returns the instance associated with that component (whether shared or not). This allows multiple components of the same type to be configured with different sets of metadata, when component operations need to be isolated from one another.
```

Author: Good suggestion. Updated.

## Registering multiple services

> Each call to `Register()` binds a socket to a registered component; one of each type of component (input/output binding, pub/sub, and state store) can be registered per socket.

Reviewer: ```suggestion
Each call to `Register()` binds a socket to a registered component. One of each component type (input/output binding, pub/sub, and state store) can be registered per socket.
```

Author: Updated.

```go
func main() {
	dapr.Register("service-a", dapr.WithStateStore(func() state.Store {
		return &components.MyDatabaseStoreComponent{}
	}))

	dapr.Register("service-a", dapr.WithOutputBinding(func() bindings.OutputBinding {
		return &components.MyDatabaseOutputBindingComponent{}
	}))

    dapr.Register("service-b", dapr.WithStateStore(func() state.Store {
		return &components.MyDatabaseStoreComponent{}
	}))

	dapr.MustRun()
}
```

## Configuring Multiple Components

Configuring Dapr to use the hosted components is the same as for any single component - the component YAML refers to the associated socket.

```yaml
#
# This component uses the state store associated with socket `service-a`
#
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: state-store-a
spec:
  type: state.service-a
  version: v1
  metadata: []
```

```yaml
#
# This component uses the state store associated with socket `service-b`
#
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: state-store-b
spec:
  type: state.service-b
  version: v1
  metadata: []

> ```

Reviewer: Could you add this to the end of this doc to help with cross-linking?

```suggestion
## Next steps

- Learn more about implementing:
  - [Bindings]({{< ref go-bindings >}})
  - [State]({{< ref go-state-store >}})
  - [Pub/sub]({{< ref go-pub-sub >}})

```

Author: Added.

---

While not typically needed by most, these guides show advanced ways you can configure your Go pluggable components.

## Component lifetime

Components are registered by passing a "factory method" that is called for each configured component of that type associated with that socket. The method returns the instance associated with that component (whether shared or not). This allows multiple components of the same type to be configured with different sets of metadata, when component operations need to be isolated from one another, etc.

## Registering multiple services

Each call to `Register()` binds a socket to a registered component. One of each component type (input/output binding, pub/sub, and state store) can be registered per socket.

```go
func main() {
	dapr.Register("service-a", dapr.WithStateStore(func() state.Store {
		return &components.MyDatabaseStoreComponent{}
	}))

	dapr.Register("service-a", dapr.WithOutputBinding(func() bindings.OutputBinding {
		return &components.MyDatabaseOutputBindingComponent{}
	}))

    dapr.Register("service-b", dapr.WithStateStore(func() state.Store {
		return &components.MyDatabaseStoreComponent{}
	}))

	dapr.MustRun()
}
```

> 

Reviewer 2: Should we make a call out that the above code defines 2 components `service-a`  and `service-b`, the first hosting 2 components while the latter hosts only 1? This would make it simpler for the reader to tie how this is linked to the 2 configurations in the section bellow and with the 2 UDS files that are created.

Author: I added some verbiage to (hopefully) help tie the two sections together.

## Configuring Multiple Components

> Configuring Dapr to use the hosted components is the same as for any single component - the component YAML refers to the associated socket. For example, to configure both state stores for the components registered above, you create two configuration files, each referencing their respective socket. 

Reviewer 2: ```suggestion
Configuring Dapr to use the hosted components is the same as for any single component - the component YAML refers to the associated socket. For example, to configure the state stores for the 2 distinct components registered above (`service-a` and `service-b`), you create two configuration files, each referencing their respective component. 
```

I also replaced _socket_ by _component_ as the code above does not mention socket -- the underlying code will create 2 sockets with distinct names but this is transparent to the programmer, no? So refering to socket here felt a bit ambiguous.

Author: It might help if I call out more clearly how the components are tied to the (socket) names in the above example.  How about:

In the example above, a state store and output binding is registered with the socket `service-a` while another state store is registered with the socket `service-b`.

## Configuring Multiple Components

Configuring Dapr to use the hosted components is the same as for any single component - the component YAML refers to the associated socket. For example, to configure Dapr state stores for the two components registered above (to sockets `service-a` and `service-b`), you create two configuration files, each referencing their respective socket.

Reviewer 2: As long as it is clear to the reader where those socket (names) are and where they come from. We know and it implicit knowledge for us - when you register the component `service-a` the SDK creates a socket named `service-a.socket` in that common directory. And this is explained in other documents that talk about pluggable components. But this particular document never mentioned how that `service-b` socket (and that is not even its name, we add a suffix, right?) came to be.
